python/cuml/ts/arima.pyx

```diff
@@ -0,0 +1,814 @@
+    handle : cumlHandle (optional)
+             The cumlHandle to be used.
+    """
+    pynvtx_range_push("ll_gf")
+    
+    fd = np.zeros(num_parameters)
+
+    grad = np.zeros(len(x))
+
+    # 1st order FD saves 20% runtime.
+    # ll_b0 = ll_f(num_batches, num_parameters, order, y, x, trans=trans)
+    assert (len(x) / num_parameters) == float(num_batches)
+    for i in range(num_parameters):
+        fd[i] = h
+
+        # duplicate the perturbation across batches (they are independent)
+        fdph = np.tile(fd, num_batches)
+
+        # reset perturbation
+        fd[i] = 0.0
+
+        ll_b_ph = ll_f(num_batches, nobs, order, y, x+fdph, trans=trans, handle=handle)
+        ll_b_mh = ll_f(num_batches, nobs, order, y, x-fdph, trans=trans, handle=handle)
+        
+        np.seterr(all='raise')
+        # first derivative second order accuracy
+        grad_i_b = (ll_b_ph - ll_b_mh)/(2*h)
+        # first derivative first order accuracy
+        # grad_i_b = (ll_b_ph - ll_b0)/(h)
+
+        if num_batches == 1:
+            grad[i] = grad_i_b
+        else:
+            assert len(grad[i::num_parameters]) == len(grad_i_b)
+            # Distribute the result to all batches
+            grad[i::num_parameters] = grad_i_b
+
+    pynvtx_range_pop()
+    return grad
+
+def fit(y,
+        order: Tuple[int, int, int],
+        mu0: np.ndarray,
+        ar_params0: List[np.ndarray],
+        ma_params0: List[np.ndarray],
+        opt_disp=-1, h=1e-9):
+    """
+    Fits an ARIMA model to each time-series for the given order and initial parameter estimates.
+
+    Parameters
+    ----------
+    y : array-like (device or host) shape = (n_samples, n_series)
+        Time series data.
+    order : Tuple[int, int, int]
+            The ARIMA order (p, d, q)
+    mu0 : array-like
+          Array of trend-parameter estimates. Only used if `d>0`.
+    ar_params0 : List of arrays
+                 List of AR parameter-arrays, one array per series, each series has `p` parameters.
+    ma_params0 : List of arrays
+                 List of MA parameter-arrays, one array per series, each series has `q` parameters.
+    opt_disp : int
+               Fit diagnostic level (for L-BFGS solver):
+               * `-1` for no output,
+               * `0<n<100` for output every `n` steps
+               * `n>100` for more detailed output
+    h        : float
+               Finite-differencing step size. The gradient is computed using second-order differencing:
+                   f(x+h) - f(x - h)
+               g = ----------------- + O(h^2)
+                         2 * h
+
+    Returns:
+    --------
+    model : ARIMAModel
+            The ARIMA model with the best fit parameters.
+
+    """
+    p, d, q = order
+    num_parameters = d + p + q
+
+    d_y, d_y_ptr, num_samples, num_batches, dtype = input_to_dev_array(y, check_dtype=np.float64)
+
+    handle = cuml.common.handle.Handle()
```

Comment:
If `fit` would be in the ARIMA class then this is inheritted from base, otherwise need to add a `handle` parameter to the `fit` method and check if its passed before creating one